fee-monitor/src/DynamicChecker.ts

```diff
@@ -0,0 +1,479 @@
+
+        const immediateSettleMoment: SettleMoment = {
+            tag: "block",
+            value: blockNumber,
+        };
+        if (this.parentTermId !== currentTermId) {
+            const blockRewardSettleMoment: SettleMoment = {
+                tag: "term",
+                value: currentTermId,
+            };
+            const totalReducedReward = await this.applyPenalty(
+                this.parentTermValidators,
+                blockNumber,
+            );
+            this.applyAdditionalRewards(this.parentTermValidators, totalReducedReward);
+            this.settleNominationDeposit(blockNumber, validators);
+            this.checkForValidators(blockRewardSettleMoment);
+            this.checkForStakeHolders(blockNumber, immediateSettleMoment, true);
+        } else {
+            this.checkForStakeHolders(blockNumber, immediateSettleMoment, false);
+        }
+        this.finalizeTerm(blockNumber, currentTermId, validators);
+    }
+
+    private async settleNominationDeposit(blockNumber: number, validators: PlatformAddress[]) {
+        const pastValidators = Array.from(this.nominationDeposits.keys());
+        const bannedAccounts = await getAccountsInState(AccountState.Banned, blockNumber);
+        const candidates = await getAccountsInState(AccountState.Candidate, blockNumber);
+        const jailedAccounts = await getAccountsInState(AccountState.Jailed, blockNumber);
+
+        const nowEligible = pastValidators.filter(
+            (address: PlatformAddress) =>
+                !bannedAccounts.includes(address) &&
+                !candidates.includes(address) &&
+                !jailedAccounts.includes(address) &&
+                !validators.includes(address),
+        );
+
+        for (const eligibleAccount of nowEligible) {
+            const deposit = this.nominationDeposits.get(eligibleAccount)!;
+            this.prevTermData.intermediateReward.deposit(eligibleAccount, deposit);
+            this.nominationDeposits.delete(eligibleAccount);
+        }
+    }
+
+    private async checkForValidators(settleMoment: SettleMoment) {
+        const errorsOfValidators: AggregatedErrors = {
+            server: SERVER,
+            moment: settleMoment,
+            errors: [] as any[],
+        };
+
+        for (const validator of this.parentTermValidators) {
+            const validatorBefore = await sdk.rpc.chain.getBalance(
+                validator,
+                this.parentTermFirstBlockNumber,
+            );
+            const validatorAfter = await sdk.rpc.chain.getBalance(
+                validator,
+                this.parentTermLastBlockNumber,
+            );
+            const settleStakeHolderReward = this.tracerForImmedateSettle.adjust(
+                validator,
+                validatorBefore,
+            );
+
+            if (
+                !this.prevTermData.intermediateReward
+                    .adjust(validator, settleStakeHolderReward)
+                    .eq(validatorAfter)
```

Comment:
The `reward` is given at the current block.

Reply:
Yes, that's a mistake. I fixed the checking routine.